tidy3d/plugins/waveguide/rectangular_dielectric.py

```diff
@@ -0,0 +1,944 @@
+        ModeSpec(num_modes=2, precision="double"),
+        title="Mode Specification",
+        description=":class:`ModeSpec` defining waveguide mode properties.",
+    )
+
+    grid_resolution: int = pydantic.Field(
+        15,
+        title="Grid Resolution",
+        description="Solver grid resolution per wavelength.",
+    )
+
+    max_grid_scaling: float = pydantic.Field(
+        1.2,
+        title="Maximal Grid Scaling",
+        description="Maximal size increase between adjacent grid boundaries.",
+    )
+
+    compute_group_index: Union[bool, pydantic.PositiveFloat] = pydantic.Field(
+        False,
+        title="Compute group index",
+        description="Control the computation of the group index alongside the effective index. If "
+        "set to a positive value, it sets the fractional frequency step used in the numerical "
+        "differentiation of the effective index to compute the group index. If set to `True`, the "
+        f"default of {GROUP_INDEX_STEP} is used.",
+    )
+
+    @pydantic.validator("wavelength", "core_width", "gap", always=True)
+    def _set_array(cls, val):
+        result = numpy.array(val, ndmin=1)
+        if any(result < 0):
+            raise ValidationError("Values may not be negative.")
+        return result
+
+    @pydantic.validator("box_medium", always=True)
+    def _set_box_medium(cls, val, values):
+        return values["clad_medium"] if val is None else val
+
+    @pydantic.validator("clad_thickness", always=True)
+    def _set_clad_thickness(cls, val, values):
+        if val is None:
+            wavelength = values["wavelength"]
+            medium = values["clad_medium"]
+            n = numpy.array([medium.nk_model(f)[0] for f in C_0 / wavelength])
+            lda = wavelength / n
+            return 1.5 * lda.max()
+        return val
+
+    @pydantic.validator("box_thickness", always=True)
+    def _set_box_thickness(cls, val, values):
+        if val is None:
+            wavelength = values["wavelength"]
+            medium = values["box_medium"]
+            n = numpy.array([medium.nk_model(f)[0] for f in C_0 / wavelength])
+            lda = wavelength / n
+            return 1.5 * lda.max()
+        return val
+
+    @pydantic.validator("side_margin", always=True)
+    def _set_side_thickness(cls, val, values):
+        return max(values["clad_thickness"], values["box_thickness"]) if val is None else val
+
+    @pydantic.validator("gap", always=True)
+    def _validate_gaps(cls, val, values):
+        if val.size == 1 and values["core_width"].size != 2:
+            return numpy.array([val[0]] * (values["core_width"].size - 1))
+        if val.size != values["core_width"].size - 1:
+            raise ValidationError("Number of gaps must be 1 less than number of core widths.")
+        return val
+
+    @pydantic.root_validator
+    def _ensure_consistency(cls, values):
+        if values["sidewall_thickness"] > 0 and values["sidewall_medium"] is None:
```

Comment:
maybe for readability, just assign `values[...]` to variables at the beginning of this.